### PR TITLE
[TTAHUB-2494] Fix for activity report approvers

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change
We (I) removed the upserts from the approval process (big mistake?)
In doing so, there was one straight up mistake and one thing that we missed.

## How to test
1/25/24 prod dataset is on keybase. Impersonate the user listed in the Jira ticket and submit report 39409.
Clear local storage; refresh the page

You will see that the report is **actually submitted**

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-2494

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/models/hooks/activityReportApprover.js [hunk 1]
```diff
@@ -41,6 +41,12 @@ const updateReportStatus = async (sequelize, instance) => {
       submissionStatus: REPORT_STATUSES.SUBMITTED,
     },
   });
+
```
Comment (thewatermethod) [inline src/models/hooks/activityReportApprover.js:44]: if you look at the query above, I changed it from 
```js
  const report = await sequelize.models.ActivityReport.findByPk(instance.activityReportId, {
    attributes: ['submissionStatus', 'calculatedStatus'],
  });
```
to
```js
  const report = await sequelize.models.ActivityReport.findByPk(instance.activityReportId, {
    attributes: ['submissionStatus', 'calculatedStatus'],
    where: {
      submissionStatus: REPORT_STATUSES.SUBMITTED,
    },
  });
```
in order to be a little more succinct
However, it should bomb out early as before the rest of the function was locked in a 
```js
if(report.submittedStatus === REPORT_STATUSES.SUBMITTED){
/// stuff
}
```

and we don't want to update a report that hasn't been submitted by a user. I don't think this is causing a bug but it is how it was working before and I don't want to introduce any new ones if I can help it

File: src/policies/activityReport.js [hunk 2]
```diff
@@ -161,8 +161,15 @@ export default class ActivityReport {
...
+  // This is a helper function to determine if the report is in a state where it can be edited
   reportHasEditableStatus() {
+    // if the report is in draft
     return this.activityReport.submissionStatus === REPORT_STATUSES.DRAFT
-      || this.activityReport.calculatedStatus === REPORT_STATUSES.NEEDS_ACTION;
+    // or if it's been marked as needs action
+      || this.activityReport.calculatedStatus === REPORT_STATUSES.NEEDS_ACTION || (
+    // or if it's submitted and the calculated status is not submitted for some reason
+      this.activityReport.submissionStatus === REPORT_STATUSES.SUBMITTED
+        && this.activityReport.calculatedStatus !== REPORT_STATUSES.SUBMITTED
```
Comment (thewatermethod) [inline src/policies/activityReport.js:172]: This is essentially damage control - the reports affected by this bug can't be edited if we don't add this little bit of business here. I can't see how it hurts anything - the report stays editable until the hooks have moved the report to "actually submitted" (see calculatedStatus) which is what we check on the frontend as well

File: src/services/activityReportApprovers.js [hunk 3]
```diff
@@ -16,11 +16,22 @@ export async function upsertApprover(values) {
     paranoid: false,
   });
 
-  if (approver && status) {
-    await approver.update({
-      status,
-      note: values.note,
-    }, {
+  if (approver) {
```
Comment (thewatermethod) [inline src/services/activityReportApprovers.js:19]: this is the actual bug. Newly submitted reports don't have an approver status. So if you
1) added an approver
2) saved
3) submitted report

You would trigger this, as the approver hooks would not refire. With this bugfix, everytime we call syncApprovers, they will ALWAYS fire

File: src/services/activityReportApprovers.js [hunk 4]
```diff
@@ -56,7 +67,7 @@ export async function upsertApprover(values) {
  * @param {*} userIds - array of userIds for approver records, ActivityReportApprovers will be
  * deleted or created to match this list
  */
-export async function syncApprovers(activityReportId, userIds = [], log = false) {
```
Comment (thewatermethod) [inline src/services/activityReportApprovers.js:59]: This is simply a bit of code I should have removed yesterday

File: src/policies/activityReport.js [hunk 5]
```diff
@@ -161,8 +161,20 @@ export default class ActivityReport {
     return approverUserIds.includes(this.user.id);
   }
 
+  // This is a helper function to determine if the report is in a state where it can be edited
   reportHasEditableStatus() {
-    return this.activityReport.submissionStatus === REPORT_STATUSES.DRAFT
-      || this.activityReport.calculatedStatus === REPORT_STATUSES.NEEDS_ACTION;
+    // if the report is in draft, it's editable
+    if (this.activityReport.submissionStatus === REPORT_STATUSES.DRAFT
+      || this.activityReport.calculatedStatus === REPORT_STATUSES.DRAFT) {
```
Comment (thewatermethod) [inline src/policies/activityReport.js:168]: This is essentially damage control - the reports affected by this bug can't be edited if we don't add this little bit of business here. I can't see how it hurts anything - the report stays editable until the hooks have moved the report to "actually submitted" (see calculatedStatus) which is what we check on the frontend as well

File: src/services/activityReportApprovers.js (hunk 3 above)
Comment (thewatermethod) [reply src/services/activityReportApprovers.js:24]: the bug being the report would then not be moved to a calculatedStatus of "submitted"

**Review by GarrettEHill — APPROVED** [review]

**Pull request merged**